docs/user-guide/examples.md

```diff
@@ -130,4 +130,35 @@ defaultEntryPoints = ["http"]
     headerField = "X-WebAuth-User"
     [entryPoints.http.auth.basic]
     users = ["test:$apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/", "test2:$apr1$d9hr9HBB$4HxwgUir3HP4EsggP/QNo0"]
+```
+
+
+## Enable authentication forwarding
+
+Uses a query param from the request to check for authentication with an auth service.
+If the auth service returns 200 OK the request is forwarded to the backend with additional headers/query params from the auth service.
+The response from the auth service is parsed, json encoded and can be forwarded to the backend as headers/query params.
+The field names in the request and the response can be manipulated for internal use via `as` config for request params 
+and the response from the auth server can be manipulated into a different name via `as` config for the response params.
+```
+defaultEntryPoints = ["http"]
+[entryPoints]
+  [entryPoints.http]
+  address = ":80"
+  [entryPoints.http.auth.forward]
+    address = "http://authserver.com/auth"
+    [entryPoints.http.auth.Forward.requestParameters.email]
+      name = "email"
+      as = "theEmail"
+    [entryPoints.http.auth.Forward.requestParameters.token]
+      name = "token"
+      as = "theToken"
+    [entryPoints.http.auth.Forward.responseReplayFields.userId]
+      path = "user.id"
+      as = "X-User-Id"
+      in = "header"
+    [entryPoints.http.auth.Forward.responseReplayFields.userName]
+      path = "user.name"
+      as = "" # No name transformation
+      in = "parameter"
```

Comment:
> I would need to force the request to be JSON but what will happen if the original request was XML will the auth server know how to handle it?

That's exactly the point. If you forward the request just like it is coming in, you are not responsible for that. The auth back end has to decide what it wants to handle. This allows the back end to provide authentication in whatever format, on whatever conditions, no matter if it's headers, body content, json or xml etc. If it gets XML and can only handle JSON it seems like the request is wrong so it should probably send a non 200 response.



> Sending random body (and I say random because there is no original route in the request) to the auth server will be without context regarding the destination of the request.

The original `Host` has to be transferred, yes. We could use `Traefik-Auth-Original-Host` for that (for example) or make it configurable in best case.

> What HTTP method should be used since currently we use GET to send the query params to send the body we would need to post it and back to point 2 if we don't use the same verb or provide this context in the request we will have a problem and the body will be useless.

The same as the original request. If it was a POST, we POST to the auth back end. If it was a GET, we GET.

Don't get me wrong, I really appreciate your work on this matter! I'm just trying to point out my ideas. The current PR is nice but it's very limiting and I'm not sure if it's a good idea to merge this and change it again later on. It will cause a lot of different configuration options whereas if we implemented it in a generic way, we basically leave it all to the auth back end. The only thing you can configure is if you want to forward the body or just the headers to save bandwidth if you don't need the body.

Side note: This is very much inspired by the way nginx has been doing it for a long time with the [ngx_http_auth_request_module](http://nginx.org/en/docs/http/ngx_http_auth_request_module.html). You can find the C source [here](https://github.com/perusio/nginx-auth-request-module/blob/master/ngx_http_auth_request_module.c).

So for me, to cover all use cases, the config should look like this and the rest should be left to the auth back end:

```toml
[entryPoints.http.auth.forward]
    endpoint = "http://authserver.com/auth"
    forwardBody = true # default: false
    [entryPoints.http.auth.forward.headerMap]
    "Host" = "Traefik-Auth-Original-Host"   
```